pkg/collector/scheduler.go

```diff
@@ -0,0 +1,190 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package collector
+
+import (
+	"expvar"
+	"fmt"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/collector/check"
+	"github.com/DataDog/datadog-agent/pkg/collector/loaders"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+var (
+	schedulerErrs *expvar.Map
+	errorStats    = newCollectorErrors()
+)
+
+func init() {
+	schedulerErrs = expvar.NewMap("CheckScheduler")
+	schedulerErrs.Set("LoaderErrors", expvar.Func(func() interface{} {
+		return errorStats.getLoaderErrors()
+	}))
+	schedulerErrs.Set("RunErrors", expvar.Func(func() interface{} {
+		return errorStats.getRunErrors()
+	}))
+}
+
+// CheckScheduler is the check scheduler
+type CheckScheduler struct {
+	configToChecks map[string][]check.ID // cache the ID of checks we load for each config
+	loaders        []check.Loader
+	collector      *Collector
+	m              sync.RWMutex
+}
+
+// NewCheckScheduler returns a check scheduler
+func NewCheckScheduler(collector *Collector) *CheckScheduler {
+	scheduler := &CheckScheduler{
+		collector:      collector,
+		configToChecks: make(map[string][]check.ID),
+		loaders:        make([]check.Loader, 0, 5),
+	}
+	// add the check loaders
+	for _, loader := range loaders.LoaderCatalog() {
+		scheduler.AddLoader(loader)
+		log.Debugf("Added %s to Check Scheduler", loader)
+	}
+	return scheduler
+}
+
+// Schedule schedules configs to checks
+func (s *CheckScheduler) Schedule(configs []integration.Config) {
+	checks := s.GetChecksFromConfigs(configs, true)
+	for _, c := range checks {
+		log.Infof("Scheduling check %s", c)
+		_, err := s.collector.RunCheck(c)
+		if err != nil {
+			log.Errorf("Unable to run Check %s: %v", c, err)
+			errorStats.setRunError(c.ID(), err.Error())
+			continue
+		}
+	}
+}
+
+// Unschedule unschedules checks matching configs
+func (s *CheckScheduler) Unschedule(configs []integration.Config) {
+	// Process removed configs first to handle the case where a
```

Comment:
I guess this comment belongs in `pollConfigs` instead?
https://github.com/DataDog/datadog-agent/pull/1936/files#diff-6f38a99c24b92894c615272d9c304d1bR390